### PR TITLE
Feat/Add object notifications

### LINKED ISSUE #1173: Produce object related notifications
Related to https://github.com/nspcc-dev/neofs-api/issues/193
Blocked by https://github.com/nspcc-dev/neofs-node/issues/1172

After new epoch event, find all objects from storage engine (shard->metabase) that should produce notification. 
Implement separate index in metabase if necessary. 
Send event using NATS JetStream client.

### PR DESCRIPTION
**Pull request by carpawell**
Blocked by https://github.com/nspcc-dev/neofs-sdk-go/pull/143 and https://github.com/nspcc-dev/neofs-api-go/pull/372

Closes #1173.

### CODE REVIEW

**Review by alexvanin — APPROVED** [review]

File: pkg/services/notificator/deps.go [hunk 1]
```diff
@@ -0,0 +1,11 @@
+package notificator
+
+import objectSDKAddress "github.com/nspcc-dev/neofs-sdk-go/object/address"
+
+// NotificationWriter is a writer that notifies all the subscribers
+// about a new object notifications.
+type NotificationWriter interface {
+	// Notify must send string representation of the object
+	// address into the specified topic.
+	Notify(topic string, address *objectSDKAddress.Address) error
```
Comment (cthulhu-rider) [inline pkg/services/notificator/deps.go:10]: I don't think `error` return currently makes sense for `Notificator`.

File: pkg/services/notificator/service.go [hunk 2]
```diff
@@ -0,0 +1,163 @@
...
+}
+
+// SetWriter sets notification writer.
+func (prm *Prm) SetWriter(v NotificationWriter) *Prm {
+	prm.writer = v
+	return prm
+}
+
+// SetStorage sets object storage.
+func (prm *Prm) SetStorage(v *engine.StorageEngine) *Prm {
```
Comment (cthulhu-rider) [inline pkg/services/notificator/service.go:42]: I would reduce the connectivity of types:
```
type NotificationSource interface {
  Iterate(epoch uint64, f(topic string, addr address.Address))  error
}

func (prm *Prm) SetNotificationSource(v NotificationSource) *Prm
```

File: pkg/services/notificator/service.go [hunk 3]
```diff
@@ -0,0 +1,163 @@
...
+	prm.storage = v
+	return prm
+}
+
+// Notificator is a notification producer that handles
+// object with defined notification epoch.
```
Comment (cthulhu-rider) [inline pkg/services/notificator/service.go:48]:
```suggestion
// objects with defined notification epoch.
```

File: pkg/services/notificator/service.go [hunk 4]
```diff
@@ -0,0 +1,163 @@
...
+
+	return &Notificator{
+		w:            prm.writer,
+		s:            prm.storage,
+		l:            prm.logger,
+		defaultTopic: *prm.defaultTopic,
+	}
+}
+
+// ProcessEpoch looks for all object with defined epoch in the storage
```
Comment (cthulhu-rider) [inline pkg/services/notificator/service.go:86]:
```suggestion
// ProcessEpoch looks for all objects with defined epoch in the storage
```

File: pkg/services/notificator/deps.go [hunk 5]
```diff
@@ -0,0 +1,11 @@
+package notificator
+
+import objectSDKAddress "github.com/nspcc-dev/neofs-sdk-go/object/address"
+
+// NotificationWriter is a writer that notifies all the subscribers
```
Comment (cthulhu-rider) [inline pkg/services/notificator/deps.go:5]:
```suggestion
// NotificationWriter notifies all the subscribers
```

File: pkg/services/notificator/deps.go [hunk 6]
```diff
@@ -0,0 +1,19 @@
+package notificator
+
+import objectSDKAddress "github.com/nspcc-dev/neofs-sdk-go/object/address"
+
+// NotificationSource provides notification that should
```
Comment (cthulhu-rider) [inline pkg/services/notificator/deps.go:5]: Just
```suggestion
// NotificationSource is a source of object notifications.
```
Interface should not be related to other interface `NotificationWriter`.

File: pkg/services/notificator/deps.go [hunk 7]
```diff
@@ -0,0 +1,19 @@
...
+// be written via the NotificationWriter.
+type NotificationSource interface {
+	// Iterate must iterate over all notification for the
```
Comment (cthulhu-rider) [inline pkg/services/notificator/deps.go:8]:
```suggestion
	// Iterate must iterate over all notifications for the
```

File: pkg/services/notificator/deps.go [hunk 8]
```diff
@@ -0,0 +1,19 @@
...
+	// provided epoch and call handler for all of them.
+	Iterate(epoch uint64, handler func(topic string, addr *objectSDKAddress.Address))
+}
+
+// NotificationWriter notifies all the subscribers
+// about a new object notifications.
```
Comment (cthulhu-rider) [inline pkg/services/notificator/deps.go:14]:
```suggestion
// about new object notifications.
```

File: pkg/services/notificator/deps.go [hunk 9]
```diff
@@ -0,0 +1,19 @@
...
+type NotificationWriter interface {
+	// Notify must send string representation of the object
```
Comment (cthulhu-rider) [inline pkg/services/notificator/deps.go:16]: Strange to see `must send string representation`, why `address` is not a `string` then?

File: pkg/services/notificator/service.go [hunk 10]
```diff
@@ -0,0 +1,86 @@
...
+
+	return &Notificator{
+		w:  prm.writer,
+		ns: prm.notificationSource,
+		l:  prm.logger,
+	}
+}
+
+// ProcessEpoch looks for all objects with defined epoch in the storage
+// and passes their addresses to the NotificationWriter, skips any erroneous
```
Comment (cthulhu-rider) [inline pkg/services/notificator/service.go:72]: Remove `skips...`

File: pkg/services/notificator/deps.go (hunk 9 above)
Comment (carpawell) [reply pkg/services/notificator/deps.go:16]: it uses part of oid to make "exactly once delivery". did not know if this should be noted in the interface description but Im sure that this part should not be parsed in the string representation

File: go.mod [hunk 11]
```diff
@@ -34,4 +34,5 @@ require (
 )
 
 // Used for debug reasons
-// replace github.com/nspcc-dev/neofs-api-go => ../neofs-api-go
+//replace github.com/nspcc-dev/neofs-api-go/v2 => ../neofs-api-go
```
Comment (cthulhu-rider) [inline go.mod:37]: Drop all the comments

**Pull request merged**